app/models/creatorship.rb

```diff
@@ -1,47 +1,278 @@
 class Creatorship < ApplicationRecord
-  belongs_to :pseud
-  belongs_to :creation, polymorphic: true, touch: true
+  belongs_to :pseud, inverse_of: :creatorships
+  belongs_to :creation, inverse_of: :creatorships, polymorphic: true, touch: true
+
+  scope :approved, -> { where(approved: true) }
+  scope :unapproved, -> { where(approved: false) }
+
+  scope :for_user, ->(user) { joins(:pseud).merge(user.pseuds) }
+
+  ########################################
+  # VALIDATIONS
+  ########################################
+
+  before_validation :update_approved, on: :create
+
+  validates_presence_of :creation
+  validates_uniqueness_of :pseud, scope: [:creation_type, :creation_id], on: :create
+
+  validate :check_invalid, on: :create
+  validate :check_disallowed, on: :create
+  validate :check_banned, on: :create
+  validate :check_approved_becoming_false, on: :update
+
+  # Update approval status if this creatorship should be automatically approved.
+  def update_approved
+    if !approved? && should_automatically_approve?
+      self.approved = true
+    end
+  end
+
+  # Make sure that the pseud exists, and isn't ambiguous.
+  #
+  # Note that thanks to the definitions of missing? and ambiguous?, this is
+  # equivalent to having a validates_presence_of :pseud check, just with
+  # different error messages.
+  def check_invalid
+    if missing?
+      errors.add(:base, ts("Could not find a pseud %{name}.", name: byline))
+    elsif ambiguous?
+      errors.add(:base, ts("The pseud %{name} is ambiguous.", name: byline))
+    end
+  end
+
+  # Make sure that if this is an invitation, we're not inviting someone who has
+  # disabled invitations.
+  def check_disallowed
+    return if approved? || pseud.nil?
+    return if pseud&.user&.preference&.allow_cocreator
+
+    errors.add(:base, ts("%{name} does not allow others to add them as a co-creator.",
```

Comment:
Since we're only ever dealing with invitations here, maybe this error should be `%{name} does not allow others to invite them to be a co-creator`.